core/services/eth/log_broadcaster_test.go

```diff
@@ -398,26 +465,213 @@ func TestLogBroadcaster_ReceivesAllLogsWhenResubscribing(t *testing.T) {
+	store, cleanup := cltest.NewStore(t)
+	defer cleanup()
+
+	const blockHeight uint64 = 0
+
+	ethClient := new(mocks.Client)
+	sub := new(mocks.Subscription)
+
+	chchRawLogs := make(chan chan<- eth.Log, 1)
+
+	ethClient.On("SubscribeToLogs", mock.Anything, mock.Anything, mock.Anything).
+		Run(func(args mock.Arguments) {
+			chRawLogs := args.Get(1).(chan<- eth.Log)
+			chchRawLogs <- chRawLogs
+		}).
+		Return(sub, nil).
+		Once()
+
+	ethClient.On("GetBlockHeight").Return(blockHeight, nil)
+	sub.On("Err").Return(nil)
+	sub.On("Unsubscribe").Return()
+
+	lb := ethsvc.NewLogBroadcaster(ethClient, store)
+	lb.Start()
+
+	job := createJob(t, store)
+	logListener := simpleLogListner{
+		func(log interface{}, err error) {},
+		*job.ID,
+	}
+	addr := common.Address{1}
+
+	lb.Register(addr, &logListener)
+
+	chRawLogs := <-chchRawLogs
+	chRawLogs <- eth.Log{Address: addr, BlockHash: cltest.NewHash(), BlockNumber: 0, Index: 0}
+	chRawLogs <- eth.Log{Address: addr, BlockHash: cltest.NewHash(), BlockNumber: 1, Index: 0}
+
+	requireLogConsumptionCount(t, store, 2)
+}
+
+func TestLogBroadcaster_ProcessesLogsOnceForEachConsumer(t *testing.T) {
+	store, cleanup := cltest.NewStore(t)
+	defer cleanup()
+
+	const blockHeight uint64 = 0
+
+	ethClient := new(mocks.Client)
+	sub := new(mocks.Subscription)
+
+	chchRawLogs := make(chan chan<- eth.Log, 1)
+
+	ethClient.On("SubscribeToLogs", mock.Anything, mock.Anything, mock.Anything).
+		Run(func(args mock.Arguments) {
+			chRawLogs := args.Get(1).(chan<- eth.Log)
+			chchRawLogs <- chRawLogs
+		}).
+		Return(sub, nil).
+		Twice()
+
+	ethClient.On("GetBlockHeight").Return(blockHeight, nil)
+	sub.On("Err").Return(nil)
+	sub.On("Unsubscribe").Return()
+
+	lb := ethsvc.NewLogBroadcaster(ethClient, store)
+	lb.Start()
+
+	job1 := createJob(t, store)
+	job2 := createJob(t, store)
+	consumedCount1 := 0
+	consumedCount2 := 0
+	logListener := simpleLogListner{
+		func(log interface{}, err error) { consumedCount1++ },
+		*job1.ID,
+	}
+	logListener2 := simpleLogListner{
+		func(log interface{}, err error) { consumedCount2++ },
+		*job2.ID,
+	}
+
+	lb.Register(common.Address{}, &logListener)
+	lb.Register(common.Address{}, &logListener2)
+
+	chRawLogs := <-chchRawLogs
+	hash := cltest.NewHash()
+	chRawLogs <- eth.Log{BlockHash: hash, BlockNumber: 0, Index: 0}
+
+	require.Eventually(t, func() bool { return consumedCount1 == 1 }, 5*time.Second, 10*time.Millisecond)
+	require.Eventually(t, func() bool { return consumedCount2 == 1 }, 5*time.Second, 10*time.Millisecond)
+	requireLogConsumptionCount(t, store, 2)
+
+	// remove last LogConsumption
+	store.ORM.RawDB(func(db *gorm.DB) error {
+		var lc models.LogConsumption
+		err := db.Where("consumer_id = ?", job2.ID).First(&lc).Error
+		require.NoError(t, err)
+		err = db.Delete(lc).Error
+		require.NoError(t, err)
+		return nil
+	})
+	requireLogConsumptionCount(t, store, 1)
+
+	chRawLogs <- eth.Log{BlockHash: hash, BlockNumber: 0, Index: 0} // same log
```

Comment:
Use a variable, instead of copying the fields? Seems less error prone if you need to add a field and keep both copies the same.